statsmodels/miscmodels/robust_genericmle.py

```diff
@@ -0,0 +1,418 @@
+                converged = True
+                break
+        else:
+            converged = False
+
+        #TODO: update res_it with iteration results
+        res_it.converged = converged
+        res_it.history = history
+
+        return res_it
+
+
+
+class RLMIterative(rlm.RLM):
+
+    def __init__(self, *args, **kwds):
+        self.weights_prior = kwds.pop('weights', None)
+        self.scale_bias = kwds.pop('weights', 1)   # TODO
+        self.meef_scale = kwds.pop('meef_scale', None)
+        self.update_scale = kwds.pop('update_scale', True)
+
+        super(RLMIterative, self).__init__(*args, **kwds)
+        # linear to get started
+        from statsmodels.regression.linear_model import WLS
+        self.fit_loc = WLS
+        self.norm_loc = self.M
+
+
+
+    def update_location(self, weight, *args):
+        pass
+
+    def fit(self, start_params=None, start_scale=None, update_scale=True,
+            optim_options=None):
+        # note scale is not squared
+        #nobs = self.endog.shape[0]
+
+        if optim_options is None:
+            optim_options = {}
+        maxiter = optim_options.get('maxiter', 10)
+        res_loc = self.fit_loc(self.endog, self.exog).fit()#, weights=None)
+        if start_params is None:
+            resid = res_loc.resid
+            params = res_loc.params
+        else:
+            params = start_params
+            resid = self.endog - res_loc.model.predict(params)
+        if start_scale is None:
+            # use MAD for now, TODO: use same as in iteration
+            scale = rscale.mad(resid, center=0)
+        else:
+            scale = start_scale
+
+        for i in range(maxiter):
+            # update location
+            weights = self.norm_loc.weights(resid / scale)
+            if self.weights_prior is not None:
+                weights *= self.weights_prior
+            res_loc = self.fit_loc(self.endog, self.exog, weights=weights).fit()
+            params_old = params
+            params = res_loc.params
+            resid = res_loc.resid
+            resid_scaled = resid / scale
+
+            if update_scale:
+                # update scale
+                # Note: use old residuals or new ?
+                weights_scale = self.meef_scale(resid_scaled) / resid_scaled**2
+                #scale_old = scale
+                scale2 = (weights_scale * resid**2).mean()
+                #scale *= scale_old**2
+                scale2 /= self.scale_bias
+                scale = np.sqrt(scale2)
+
+            #check convergence
+            if np.allclose(params, params_old, rtol=1e-6, atol=1e-10):
+                converged = True
+                break
+        else:
+            converged = False
+
+        # Force estimated scale for WLS results
+        res_loc._cache['scale'] = scale**2
+
+        # TODO return Results
+        return res_loc, scale, converged
+
+
+class _RobustGMM(gmm.GMM):
+    """Experimental class for robust M-estimation using estimating equations
```

Comment:
this is not generic GMM, for arbitrary moment conditions.
This assumes additive residual = y - predict.